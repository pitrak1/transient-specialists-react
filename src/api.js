import axios from 'axios'
import utils from './utils'

const attachInterceptors = () => {
  axios.interceptors.response.use(
    function(response) {
      return response.data.statusCode === 500
        ? Promise.reject(response)
        : response
    },
    function(error) {
      return Promise.reject(error)
    },
  )
}

const translateError = error => {
  if (error.includes('duplicate key value violates unique constraint')) {
    if (error.includes('idx_type_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_oem_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_model_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_equipment_serial_number')) {
      return 'Serial Number must be unique (case insensitive)'
    }
  }

  if (error.includes('violates foreign key constraint')) {
    if (error.includes('equipments_model_id_fkey')) {
      return 'Model cannot be deleted because it has related equipment'
    }

    if (error.includes('equipments_type_id_fkey')) {
      return 'Type cannot be deleted because it has related equipment'
    }

    if (error.includes('models_oem_id_fkey')) {
      return 'OEM cannot be deleted because it has related models'
    }
  }

  return error
}

const getIndex = (resource, options, success, failure) => {
  attachInterceptors()
  const { ascending, page, perPage, searchValue, sortBy } = options
  return axios
    .get(
      `${
        process.env.LAMBDA_ENDPOINT
      }${resource}?ascending=${ascending}&page=${page}&perPage=${perPage}&searchValue=${searchValue}&sortBy=${utils.pascalToSnake(
        sortBy,
      )}`,
    )
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

const getShow = (resource, id, options, success, failure) => {
  attachInterceptors()
  const { ascending, page, perPage, sortBy } = options
  return axios
    .get(
      `${
        process.env.LAMBDA_ENDPOINT
      }${resource}?id=${id}&ascending=${ascending}&page=${page}&perPage=${perPage}&sortBy=${utils.pascalToSnake(
        sortBy,
      )}`,
    )
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

const getNew = (resource, success, failure) => {
  attachInterceptors()
  return axios
    .get(`${process.env.LAMBDA_ENDPOINT}${resource}?new=true`)
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

export const postCreate = (resource, data, success, failure) => {
  attachInterceptors()
  return axios
    .post(`${process.env.LAMBDA_ENDPOINT}${resource}`, data)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export const patchUpdate = (resource, data, success, failure) => {
  attachInterceptors()
  return axios
    .patch(`${process.env.LAMBDA_ENDPOINT}${resource}`, data)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export const deleteDestroy = (resource, id, success, failure) => {
  attachInterceptors()
  return axios
    .delete(`${process.env.LAMBDA_ENDPOINT}${resource}?id=${id}`)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export default {
  getIndex,
  getShow,
  getNew,
  postCreate,
  patchUpdate,
  deleteDestroy,
}
