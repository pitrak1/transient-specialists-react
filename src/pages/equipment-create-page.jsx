import React from 'react'
import { Spinner } from '@instructure/ui-elements'
import { Alert } from '@instructure/ui-alerts'
import { Heading } from '@instructure/ui-elements'
import { Button } from '@instructure/ui-buttons'
import FormTextInput from '../components/form-text-input.jsx'
import FormSelect from '../components/form-select.jsx'
import api from '../api.js'
import { withRouter } from 'react-router'

class EquipmentCreatePage extends React.Component {
  state = {
    alert: null,
    data: {},
    error: null,
    loading: true,
    modelId: null,
    modelIdValid: false,
    oemId: null,
    oemIdValid: false,
    serialNumber: '',
    serialNumberValid: false,
    typeId: null,
    typeIdValid: false,
  }

  componentDidMount() {
    api.getNew(
      'equipment',
      result => {
        this.setState({ loading: false, data: result })
      },
      error => {
        this.setState({ loading: false, error })
      },
    )
  }

  handleChange = (identifier, value, valid) => {
    const state = {}
    state[identifier] = value
    state[`${identifier}Valid`] = valid

    if (identifier === 'oemId') {
      state.modelId = null
      state.modelIdValid = false
    }

    this.setState(state)
  }

  handleClick = () => {
    this.setState({ loading: true, alert: null })
    api.postCreate(
      'equipment',
      {
        serialNumber: this.state.serialNumber,
        modelId: this.state.modelId,
        typeId: this.state.typeId,
      },
      () => {
        this.props.history.push('/')
      },
      error => {
        this.setState({ loading: false, alert: error })
      },
    )
  }

  render() {
    if (this.state.loading) {
      return <Spinner renderTitle='Loading' size='large' />
    }

    if (this.state.error) {
      return <Alert variant='error'>{this.state.error}</Alert>
    }

    return (
      <div>
        {this.state.alert && <Alert variant='error'>{this.state.alert}</Alert>}
        <Heading level='h1' margin='medium'>
          Add Equipment
        </Heading>
        <FormTextInput
          identifier='serialNumber'
          label='Serial Number'
          onChange={this.handleChange}
          required={true}
          value={this.state.serialNumber}
        />
        <FormSelect
          defaultOptionLabel='Select an OEM'
          disabled={false}
          identifier={'oemId'}
          label='OEM'
          onChange={this.handleChange}
          options={this.state.data.oems}
          required={true}
          value={this.state.oemId}
        />
        <FormSelect
          defaultOptionLabel='Select a Model'
          disabled={!this.state.oemId}
          identifier={'modelId'}
          label='Model'
          onChange={this.handleChange}
          options={this.state.data.models.filter(
            model => model.oemId === this.state.oemId,
          )}
          required={true}
          value={this.state.modelId}
        />
        <FormSelect
          defaultOptionLabel='Select a Type'
          disabled={false}
          identifier={'typeId'}
          label='Type'
          onChange={this.handleChange}
          options={this.state.data.types}
          required={true}
          value={this.state.typeId}
        />
        <Button
          disabled={
            !(
              this.state.serialNumberValid &&
              this.state.oemIdValid &&
              this.state.modelIdValid &&
              this.state.typeIdValid
            )
          }
          onClick={this.handleClick}
        >
          Submit
        </Button>
      </div>
    )
  }
}

export default withRouter(EquipmentCreatePage)
