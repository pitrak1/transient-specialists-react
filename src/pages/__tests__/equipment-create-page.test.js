import React from 'react'
import { EquipmentCreatePage } from '../equipment-create-page'
import { shallow } from 'enzyme'
import sinon from 'sinon'
import api from '../../api.js'

describe('EquipmentCreatePage', () => {
  const data = {
    oems: [
      { id: 1, name: 'OEM 1' },
      { id: 2, name: 'OEM 2' },
      { id: 3, name: 'OEM 3' },
    ],
    models: [
      { id: 4, name: 'Model 1', oemId: 1 },
      { id: 5, name: 'Model 2', oemId: 2 },
      { id: 6, name: 'Model 3', oemId: 3 },
    ],
    types: [
      { id: 7, name: 'Type 1' },
      { id: 8, name: 'Type 2' },
      { id: 9, name: 'Type 3' },
    ],
  }

  it('renders spinner if not yet loaded', () => {
    const node = shallow(<EquipmentCreatePage />)
    expect(node).toMatchSnapshot()
  })

  it('renders an error if load fails', () => {
    api.getNew = sinon.stub().callsArgWith(2, 'some error')
    const node = shallow(<EquipmentCreatePage />)
    expect(node).toMatchSnapshot()
  })

  it('renders form if load succeeds', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    const node = shallow(<EquipmentCreatePage />)
    expect(node).toMatchSnapshot()
  })

  it('enables model select if OEM ID is set', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    const node = shallow(<EquipmentCreatePage />)
    node.instance().handleChange('oemId', 1, true)
    expect(node).toMatchSnapshot()
  })

  it('enables submit if fields are valid', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    const node = shallow(<EquipmentCreatePage />)
    node.instance().handleChange('serialNumber', 'Equipment 1', true)
    node.instance().handleChange('oemId', 1, true)
    node.instance().handleChange('modelId', 4, true)
    node.instance().handleChange('typeId', 7, true)
    expect(node).toMatchSnapshot()
  })

  it('redirects to equipment if submit succeeds', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    api.postCreate = sinon.stub().callsArg(2)
    const push = sinon.stub()
    const node = shallow(<EquipmentCreatePage history={{ push }} />)
    node.instance().handleClick()
    expect(push.firstCall.args[0]).toBe('/')
  })

  it('displays alert if submit fails', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    api.postCreate = sinon.stub().callsArgWith(3, 'some error')
    const node = shallow(<EquipmentCreatePage />)
    node.instance().handleClick()
    expect(node).toMatchSnapshot()
  })
})
