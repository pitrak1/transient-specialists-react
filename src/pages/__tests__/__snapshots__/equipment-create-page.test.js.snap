// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EquipmentCreatePage displays alert if submit fails 1`] = `
<div>
  <Alert
    liveRegionPoliteness="assertive"
    margin="x-small 0"
    open={true}
    screenReaderOnly={false}
    timeout={0}
    transition="fade"
    variant="error"
  >
    some error
  </Alert>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    Add Equipment
  </Heading>
  <FormTextInput
    identifier="serialNumber"
    label="Serial Number"
    onChange={[Function]}
    required={true}
    value=""
  />
  <FormSelect
    defaultOptionLabel="Select an OEM"
    disabled={false}
    identifier="oemId"
    label="OEM"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 1,
          "name": "OEM 1",
        },
        Object {
          "id": 2,
          "name": "OEM 2",
        },
        Object {
          "id": 3,
          "name": "OEM 3",
        },
      ]
    }
    required={true}
    value={null}
  />
  <FormSelect
    defaultOptionLabel="Select a Model"
    disabled={true}
    identifier="modelId"
    label="Model"
    onChange={[Function]}
    options={Array []}
    required={true}
    value={null}
  />
  <FormSelect
    defaultOptionLabel="Select a Type"
    disabled={false}
    identifier="typeId"
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 7,
          "name": "Type 1",
        },
        Object {
          "id": 8,
          "name": "Type 2",
        },
        Object {
          "id": 9,
          "name": "Type 3",
        },
      ]
    }
    required={true}
    value={null}
  />
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    disabled={true}
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Submit
  </Button>
</div>
`;

exports[`EquipmentCreatePage enables model select if OEM ID is set 1`] = `
<div>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    Add Equipment
  </Heading>
  <FormTextInput
    identifier="serialNumber"
    label="Serial Number"
    onChange={[Function]}
    required={true}
    value=""
  />
  <FormSelect
    defaultOptionLabel="Select an OEM"
    disabled={false}
    identifier="oemId"
    label="OEM"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 1,
          "name": "OEM 1",
        },
        Object {
          "id": 2,
          "name": "OEM 2",
        },
        Object {
          "id": 3,
          "name": "OEM 3",
        },
      ]
    }
    required={true}
    value={1}
  />
  <FormSelect
    defaultOptionLabel="Select a Model"
    disabled={false}
    identifier="modelId"
    label="Model"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 4,
          "name": "Model 1",
          "oemId": 1,
        },
      ]
    }
    required={true}
    value={null}
  />
  <FormSelect
    defaultOptionLabel="Select a Type"
    disabled={false}
    identifier="typeId"
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 7,
          "name": "Type 1",
        },
        Object {
          "id": 8,
          "name": "Type 2",
        },
        Object {
          "id": 9,
          "name": "Type 3",
        },
      ]
    }
    required={true}
    value={null}
  />
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    disabled={true}
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Submit
  </Button>
</div>
`;

exports[`EquipmentCreatePage enables submit if fields are valid 1`] = `
<div>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    Add Equipment
  </Heading>
  <FormTextInput
    identifier="serialNumber"
    label="Serial Number"
    onChange={[Function]}
    required={true}
    value="Equipment 1"
  />
  <FormSelect
    defaultOptionLabel="Select an OEM"
    disabled={false}
    identifier="oemId"
    label="OEM"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 1,
          "name": "OEM 1",
        },
        Object {
          "id": 2,
          "name": "OEM 2",
        },
        Object {
          "id": 3,
          "name": "OEM 3",
        },
      ]
    }
    required={true}
    value={1}
  />
  <FormSelect
    defaultOptionLabel="Select a Model"
    disabled={false}
    identifier="modelId"
    label="Model"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 4,
          "name": "Model 1",
          "oemId": 1,
        },
      ]
    }
    required={true}
    value={4}
  />
  <FormSelect
    defaultOptionLabel="Select a Type"
    disabled={false}
    identifier="typeId"
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 7,
          "name": "Type 1",
        },
        Object {
          "id": 8,
          "name": "Type 2",
        },
        Object {
          "id": 9,
          "name": "Type 3",
        },
      ]
    }
    required={true}
    value={7}
  />
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    disabled={false}
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Submit
  </Button>
</div>
`;

exports[`EquipmentCreatePage renders an error if load fails 1`] = `
<Alert
  liveRegionPoliteness="assertive"
  margin="x-small 0"
  open={true}
  screenReaderOnly={false}
  timeout={0}
  transition="fade"
  variant="error"
>
  some error
</Alert>
`;

exports[`EquipmentCreatePage renders form if load succeeds 1`] = `
<div>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    Add Equipment
  </Heading>
  <FormTextInput
    identifier="serialNumber"
    label="Serial Number"
    onChange={[Function]}
    required={true}
    value=""
  />
  <FormSelect
    defaultOptionLabel="Select an OEM"
    disabled={false}
    identifier="oemId"
    label="OEM"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 1,
          "name": "OEM 1",
        },
        Object {
          "id": 2,
          "name": "OEM 2",
        },
        Object {
          "id": 3,
          "name": "OEM 3",
        },
      ]
    }
    required={true}
    value={null}
  />
  <FormSelect
    defaultOptionLabel="Select a Model"
    disabled={true}
    identifier="modelId"
    label="Model"
    onChange={[Function]}
    options={Array []}
    required={true}
    value={null}
  />
  <FormSelect
    defaultOptionLabel="Select a Type"
    disabled={false}
    identifier="typeId"
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 7,
          "name": "Type 1",
        },
        Object {
          "id": 8,
          "name": "Type 2",
        },
        Object {
          "id": 9,
          "name": "Type 3",
        },
      ]
    }
    required={true}
    value={null}
  />
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    disabled={true}
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Submit
  </Button>
</div>
`;

exports[`EquipmentCreatePage renders spinner if not yet loaded 1`] = `
<Spinner
  as="div"
  renderTitle="Loading"
  size="large"
  variant="default"
/>
`;
